### PR TITLE
Scheduler refactor

### PR DESCRIPTION
**Pull request by bearpelican**
I believe most of the bugs are worked out. 
Will need to do a full run tonight to make sure everything converges

### CODE REVIEW

File: pytorch/training/resnet.py [hunk 1]
```diff
@@ -35,208 +17,121 @@ def forward(self, x): return x.view(x.size(0), -1)
...
-
         out = self.conv2(out)
-        out = self.bn2(out)
-        out = self.relu(out)
 
         out = self.conv3(out)
         out = self.bn3(out)
 
+        if self.downsample is not None:
+            residual = self.downsample(x)
```
[the remark this answers is not in the document]
Comment (bearpelican) [reply pytorch/training/resnet.py]: This is basically the exact code from torchvision: https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py

I think downsample is just a conv layer (passed in) with stride of 2. I think they pass it in because not all the blocks need it

File: pytorch/launch_nv.py [hunk 2]
```diff
@@ -85,157 +85,200 @@
...
+lr = 4e-1
 x_args = [
-  '--lr-sched', '4,20,34,42',
-  '--batch-sched', '192,192,128',
-  '--resize-sched', '0,18,41',
-  '--epochs', 45,
-  '--lr', 0.4,
-  '--dist-url', 'file:///home/ubuntu/data/file.sync', # single instances are faster with file sync
+  '--lr-phases', [
+    {'ep':0,  'sz':128, 'bs':192, 'trndir':'-sz/160'},
```
[the remark this answers is not in the document]
Comment (bearpelican) [reply pytorch/launch_nv.py]: Yup you are right. Not setting trndir means we use the original size image folder. I'll make it more explicit on what's happening. 

There's now a flag that let's you reuse the dataloader if the size doesn't change - 'keep_dl':True

File: pytorch/launch_nv.py [hunk 3]
```diff
@@ -85,157 +85,200 @@
...
-  '--no-bn-wd',
-  '--c10d'
+  '--dist-url', 'file:///home/ubuntu/data/file.sync', # single instances are faster with file sync
+  # '--c10d'
 ]
 
-# 2x p3's - is this still being used?
+
+# 2x p3 for throughput benchmarking
+lr = 1.0
```
[the remark this answers is not in the document]
Comment (bearpelican) [reply pytorch/launch_nv.py]: I've 1.5x'd the batch size here, so I scaled the learning rate accordingly. Sorry, I didn't know that we were still using this, otherwise I wouldn't have changed anything. Feel free to revert these changes

[the remark this answers is not in the document]
Comment (yaroslavvb) [reply pytorch/launch_nv.py]: ah ok, larger lr makes sense then

File: pytorch/training/train_imagenet_nv.py [hunk 4]
```diff
@@ -125,127 +117,123 @@ def get_parser():
...
+            if previous_bs: phase['autoscale_lr'] = phase['bs']/previous_bs
+            previous_bs = phase['bs']
 
-    def load_data(self, dir_prefix, valdir_prefix, batch_size, image_size, **kwargs):
+        return phases
+
+    def expand_directories(self, phase):
+        trndir = phase.get('trndir', '')
+        valdir = phase.get('valdir', trndir)
+        phase['trndir'] = args.data+trndir+'/train'
```
[the remark this answers is not in the document]
Comment (bearpelican) [reply pytorch/training/train_imagenet_nv.py]: trndir is really the folder extension as we switch between '~/data/imagenet' or '~/data/imagenet-sz/160'

So we actually don't want the / in this case. We'll have to figure out a more clear way get the correct folder

**Comment by bearpelican** [comment]
Merging this for now so you can run the tests. Will revisit some of the comments later

File: pytorch/training/train_imagenet_nv.py [hunk 5]
```diff
@@ -353,10 +342,8 @@ def main():
         best_prec5 = max(prec5, best_prec5)
         if args.local_rank == 0:
             if is_best: save_checkpoint(epoch, model, best_prec5, optimizer, is_best=True, filename='model_best.pth.tar')
-            if (epoch+1)==int(args.epochs*dm.resize_sched[0]+0.5):
-                save_checkpoint(epoch, model, best_prec5, optimizer, filename='sz128_checkpoint.path.tar')
-            elif (epoch+1)==int(args.epochs*dm.resize_sched[1]+0.5):
-                save_checkpoint(epoch, model, best_prec5, optimizer, filename='sz244_checkpoint.path.tar')
+            phase = dm.get_phase(epoch)
+            if phase:save_checkpoint(epoch, model, best_prec5, optimizer, filename=f'sz{phase["bs"]}_checkpoint.path.tar')
```
[the remark this answers is not in the document]
Comment (bearpelican) [reply pytorch/training/train_imagenet_nv.py]: Yeah - this will be a larger task on the list https://github.com/diux-dev/cluster/issues/29

File: pytorch/training/resnet_arch_tests.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
+import torch.nn as nn
```
[the remark this answers is not in the document]
Comment (bearpelican) [reply pytorch/training/resnet_arch_tests.py]: Deleted! It was our old resnet.py with different tests

File: pytorch/training/train_imagenet_nv.py [hunk 7]
```diff
@@ -65,28 +65,21 @@ def __getattr__(self, *args):
...
-                        metavar='N', help='mini-batch size (default: 256)')
     parser.add_argument('--resize-sched', default='0,18,41', type=str,
                         help='Scheduler to resize from 128 -> 224 -> 288')
-    parser.add_argument('--use-352-folder', action='store_true', help='Train images from 352 resized folder - faster at cost of accuracy')
-    parser.add_argument('--lr-sched', default='5,21,35,43', type=str,
-                        help='Learning rate scheduler warmup -> lr -> lr/10 -> lr/100 -> lr/1000')
-    parser.add_argument('--lr-linear-scale', action='store_true',
-                        help='Linear scale the learning rate if we change the batch size later on')
+    parser.add_argument('--no-autoscale-lr2batch', action='store_true',
+                        help='Currently automatically scaling the learning rate if the batch size changes midway through training. Enable flag to turn this off')
```
[the remark this answers is not in the document]
Comment (bearpelican) [reply pytorch/training/train_imagenet_nv.py]: Yeah I'll make the default False so it's more clear what is going on. Before the refactor and new training schedule, this was necessary
